new SessionControl library (#1)
beakerlib/ControlFlow: SessionControl/README.md

# NAME

**library(ses/basic)**

# DESCRIPTION

A library providing functions to support multiple sessions control.

# VARIABLES

- **sesID**

    An array holding currently open session IDs. Sessions are strored from index 1,
    index 0 is always used for the "default" **ID**. The default **ID** is always reset
    to the last used _ID_.

> - **sesRunTimeout**

Reviewer: Shall the variables be in uppercase letters? At the first sight this a bit confusing to me.

Author: right, probably `sesRunTIMEOUT` might be better

## sesExpect

Similarly to an `expect` script, wait for a _REG\_EXP_ pattern appearence
in the **sesID\[0\]** session.

    sesExpect [options] REG_EXP

### options

- **--id** _ID_

    If provided the **sesID\[0\]** will be set to _ID_.

- **--timeout** _TIMEOUT_

    The command execution time will be limitted to _TIMEOUT_ second(s).

    Defaults to **120** seconds or **sesExpectTimeout**, if set.

- _REG\_EXP_

    The pattern to be awaited in the session output.

Both _STDOUT_ and _STDERR_ of the session will be merged and passed to
_STDOUT_ continuously.

Returns **0** if successful. See section ["COMMON RESULT CODE"](#common-result-code) for more details.

## sesSend

Similarly to an `expect` script, send an _INPUT_ to the **sesID\[0\]** session.

    sesSend [options] INPUT

### options

- **--id** _ID_

    If provided the **sesID\[0\]** will be set to _ID_.

- _INPUT_

    The input to be send to the session. It may contain also control characters,
    e.g. **\\003** to send break (^C).

    Note, to execute a command using `sesSend` you need to append **\\r** to confirm
    it on the prompt.

Returns **0** if successful.

> ## sesSend

Reviewer: ```suggestion
## sesRaw
```

Author: good catch

Send raw expect code the session handling daemon.

    sesRaw [options] CODE

### options

- **--id** _ID_

    If provided the **sesID\[0\]** will be set to _ID_.

- _CODE_

    The code to be executed in the session handling expect daemon.

    If `-` is passed, the code will be read from _STDIN_.

Returns **0** if successful. See section ["COMMON RESULT CODE"](#common-result-code) for more details.

## sesClose

Close the opened session **sesID\[0\]**.

    sesClose [options]

### options

- **--id** _ID_

    If provided the **sesID\[0\]** will be set to _ID_.

Returns **0** if successful. See section ["COMMON RESULT CODE"](#common-result-code) for more details.

## sesCleanup

Close all the remaining open sessions.

    sesCleanup

Returns **0** if successful. See section ["COMMON RESULT CODE"](#common-result-code) for more details.

# COMMON RESULT CODE

> There are specail _RETURN CODES_ commning from the library's functions.

Reviewer: ```suggestion
There are special _RETURN CODES_ coming from the library's functions.
```

- **255**

    The session ended unexpectedly. Any further interactions with the session will
    end with an error.

- **254**

    The session call timed out. There may be a _command_ hanging or a _pattern_ was
    not found in time.

- **<254**

>     These return codes are typically comming from the executed command.

Reviewer: ```suggestion
    These return codes are typically coming from the executed command.
```

- **0**

    Success!

# EXAMPLES

Simply run `whoami` command in a session

    sesOpen
    sesRun "id"

>     ses Close

Reviewer: ```suggestion
    sesClose
```
